widget/gridwraplist.go

```diff
@@ -0,0 +1,355 @@
+	"fyne.io/fyne/v2/data/binding"
+	"fyne.io/fyne/v2/theme"
+	"fyne.io/fyne/v2/widget"
+)
+
+// Declare conformity with Widget interface.
+var _ fyne.Widget = (*GridWrapList)(nil)
+
+// GridWrapListItemID is the ID of an individual item in the GridWrapList widget.
+type GridWrapListItemID int
+
+// GridWrapList is a widget with an API very similar to widget.List,
+// that lays out items in a wrapping grid similar to container.NewGridWrap.
+// It caches and reuses widgets for performance.
+type GridWrapList struct {
+	widget.BaseWidget
+
+	Length     func() int                                          `json:"-"`
+	CreateItem func() fyne.CanvasObject                            `json:"-"`
+	UpdateItem func(id GridWrapListItemID, item fyne.CanvasObject) `json:"-"`
+
+	scroller      *container.Scroll
+	itemMin       fyne.Size
+	offsetY       float32
+	offsetUpdated func(fyne.Position)
+}
+
+// NewGridWrapList creates and returns a GridWrapList widget for displaying items in
+// a wrapping grid layout with scrolling and caching for performance.
+func NewGridWrapList(length func() int, createItem func() fyne.CanvasObject, updateItem func(GridWrapListItemID, fyne.CanvasObject)) *GridWrapList {
+	gwList := &GridWrapList{BaseWidget: widget.BaseWidget{}, Length: length, CreateItem: createItem, UpdateItem: updateItem}
+	gwList.ExtendBaseWidget(gwList)
+	return gwList
+}
+
+// NewGridWrapListWithData creates a new GridWrapList widget that will display the contents of the provided data.
+func NewGridWrapListWithData(data binding.DataList, createItem func() fyne.CanvasObject, updateItem func(binding.DataItem, fyne.CanvasObject)) *GridWrapList {
+	gwList := NewGridWrapList(
+		data.Length,
+		createItem,
+		func(i GridWrapListItemID, o fyne.CanvasObject) {
+			item, err := data.GetItem(int(i))
+			if err != nil {
+				fyne.LogError(fmt.Sprintf("Error getting data item %d", i), err)
+				return
+			}
+			updateItem(item, o)
+		})
+
+	data.AddListener(binding.NewDataListener(gwList.Refresh))
+	return gwList
+}
+
+// CreateRenderer is a private method to Fyne which links this widget to its renderer.
+func (l *GridWrapList) CreateRenderer() fyne.WidgetRenderer {
+	l.ExtendBaseWidget(l)
+
+	if f := l.CreateItem; f != nil {
+		if l.itemMin.IsZero() {
+			l.itemMin = f().MinSize()
+		}
+	}
+	layout := &fyne.Container{}
+	l.scroller = container.NewVScroll(layout)
+	layout.Layout = newGridWrapListLayout(l)
+	layout.Resize(layout.MinSize())
+	objects := []fyne.CanvasObject{l.scroller}
+	lr := newGridWrapListRenderer(objects, l, l.scroller, layout)
+	return lr
+}
+
+// MinSize returns the size that this widget should not shrink below.
+func (l *GridWrapList) MinSize() fyne.Size {
+	l.ExtendBaseWidget(l)
+
+	return l.BaseWidget.MinSize()
+}
+
+func (l *GridWrapList) scrollTo(id GridWrapListItemID) {
+	if l.scroller == nil {
+		return
+	}
+	row := math.Floor(float64(id) / float64(l.getColCount()))
+	y := (float32(row) * l.itemMin.Height) + (float32(row) * theme.Padding())
```

Comment:
Not a big deal but removing the parentheses makes it slightly easier to read in my opinion (it is of course logically the same per the operation order rules).
```suggestion
	y := float32(row) * l.itemMin.Height + float32(row) * theme.Padding()
```